File: examples/python/query_graph.py
#!/usr/bin/env python3
"""Inspect the checked-in BTU Graph snapshot without third-party packages."""

import json
from collections import Counter
from pathlib import Path


graph_path = Path(__file__).resolve().parents[2] / "data" / "graph.json"
graph = json.loads(graph_path.read_text(encoding="utf-8"))
nodes = {node["id"]: node for node in graph["nodes"]}

for edge in graph["edges"]:
    if edge["from"] not in nodes or edge["to"] not in nodes:
        raise ValueError(f"Broken edge: {edge['from']} -> {edge['to']}")

print(f"BTU Graph edition {graph['edition']}")
print(f"{len(nodes)} nodes, {len(graph['edges'])} relationships")

for node_type, count in sorted(Counter(node["type"] for node in nodes.values()).items()):
    print(f"{node_type}: {count}")

print("\nCompany-to-software relationships:")
product_links = sorted(
    f"{nodes[edge['from']]['name']} -> {nodes[edge['to']]['name']}"
    for edge in graph["edges"]
    if edge["relationship"] == "offers_software_product"
)
print("\n".join(product_links))

File: examples/javascript/query-graph.mjs
import { readFile } from "node:fs/promises";

const graphUrl = new URL("../../data/graph.json", import.meta.url);
const graph = JSON.parse(await readFile(graphUrl, "utf8"));
const nodes = new Map(graph.nodes.map((node) => [node.id, node]));

for (const edge of graph.edges) {
  if (!nodes.has(edge.from) || !nodes.has(edge.to)) {
    throw new Error(`Broken edge: ${edge.from} -> ${edge.to}`);
  }
}

const counts = Object.groupBy
  ? Object.groupBy(graph.nodes, (node) => node.type)
  : graph.nodes.reduce((groups, node) => {
      (groups[node.type] ??= []).push(node);
      return groups;
    }, {});

console.log(`BTU Graph edition ${graph.edition}`);
console.log(`${graph.nodes.length} nodes, ${graph.edges.length} relationships`);
console.log(
  Object.entries(counts)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([type, items]) => `${type}: ${items.length}`)
    .join("\n"),
);

const productLinks = graph.edges
  .filter((edge) => edge.relationship === "offers_software_product")
  .map((edge) => `${nodes.get(edge.from).name} -> ${nodes.get(edge.to).name}`)
  .sort();

console.log("\nCompany-to-software relationships:");
console.log(productLinks.join("\n"));

